src/Grpc.Net.Client/Internal/Configuration/ConfigProperty.cs

```diff
@@ -37,13 +37,23 @@ public ConfigProperty(Func<TInner?, TValue?> valueFactory, string key)
         {
             if (_value == null)
             {
-                var innerValue = inner.GetValue<TInner>(_key);
-                _value = _valueFactory(innerValue);
-
-                if (_value != null && innerValue == null)
+                // Multiple threads can get a property at the same time. We want this to be safe.
+                // Because a value could be lazyily initialized, lock to ensure multiple threads
+                // don't try to update the underlying dictionary at the same time.
+                lock (this)
                 {
-                    // Set newly created value
-                    SetValue(inner, _value);
+                    // Double-check locking.
+                    if (_value == null)
+                    {
+                        var innerValue = inner.GetValue<TInner>(_key);
+                        _value = _valueFactory(innerValue);
```

Comment:
I probably should have asked this before approving, but can `_valueFactory` call into user code? Or `inner.GetValue/SetValue`?

Reply:
No, it's internal.